File: php/laravel/app/Repositories/PostgresPetstoreRepository.php
<?php

declare(strict_types=1);

namespace App\Repositories;

use PDO;
use PDOException;

class PostgresPetstoreRepository implements PetstoreRepositoryInterface
{
    private PDO $pdo;

    public function __construct()
    {
        $host     = env('DB_HOST', 'localhost');
        $port     = env('DB_PORT', '5434');
        $database = env('DB_DATABASE', 'php-laravel');
        $username = env('DB_USERNAME', 'myuser');
        $password = env('DB_PASSWORD', 'mypassword');

        $dsn = "pgsql:host={$host};port={$port};dbname={$database}";
        $this->pdo = new PDO($dsn, $username, $password, [
            PDO::ATTR_ERRMODE            => PDO::ERRMODE_EXCEPTION,
            PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
        ]);
    }

    // ── Pet ──────────────────────────────────────────────────────────────

    public function addPet(array $pet): array
    {
        $id        = $pet['id'] ?? null;
        $name      = $pet['name'] ?? '';
        $status    = $pet['status'] ?? 'available';
        $category  = isset($pet['category']) ? json_encode($pet['category']) : null;
        $photoUrls = json_encode($pet['photoUrls'] ?? []);
        $tags      = json_encode($pet['tags'] ?? []);

        if ($id === null || $id === 0) {
            $stmt = $this->pdo->query('SELECT COALESCE(MAX(id),0)+1 AS next_id FROM pet');
            $id   = (int) $stmt->fetchColumn();
        }

        $sql = <<<SQL
            INSERT INTO pet (id, name, status, category, photo_urls, tags)
            VALUES (:id, :name, :status::pet_status, :category, :photo_urls, :tags)
            ON CONFLICT (id) DO UPDATE
                SET name       = EXCLUDED.name,
                    status     = EXCLUDED.status,
                    category   = EXCLUDED.category,
                    photo_urls = EXCLUDED.photo_urls,
                    tags       = EXCLUDED.tags
            RETURNING id, name, status::text, category, photo_urls, tags
        SQL;

        $stmt = $this->pdo->prepare($sql);
        $stmt->execute([
            ':id'         => $id,
            ':name'       => $name,
            ':status'     => $status,
            ':category'   => $category,
            ':photo_urls' => $photoUrls,
            ':tags'       => $tags,
        ]);

        return $this->hydratePet($stmt->fetch());
    }

    public function updatePet(array $pet): ?array
    {
        if (empty($pet['id'])) {
            return null;
        }
        return $this->addPet($pet);
    }

    public function findPetsByStatus(string $status): array
    {
        $stmt = $this->pdo->prepare(
            'SELECT id, name, status::text, category, photo_urls, tags FROM pet WHERE status = :status::pet_status'
        );
        $stmt->execute([':status' => $status]);
        return array_map([$this, 'hydratePet'], $stmt->fetchAll());
    }

    public function findPetsByTags(array $tags): array
    {
        if (empty($tags)) {
            $stmt = $this->pdo->query('SELECT id, name, status::text, category, photo_urls, tags FROM pet');
            return array_map([$this, 'hydratePet'], $stmt->fetchAll());
        }

        $placeholders = implode(',', array_fill(0, count($tags), '?'));
        $sql = <<<SQL
            SELECT id, name, status::text, category, photo_urls, tags
            FROM pet, jsonb_array_elements(tags::jsonb) AS t
            WHERE t->>'name' IN ({$placeholders})
            GROUP BY id
        SQL;
        $stmt = $this->pdo->prepare($sql);
        $stmt->execute($tags);
        return array_map([$this, 'hydratePet'], $stmt->fetchAll());
    }

    public function getPetById(int $petId): ?array
    {
        $stmt = $this->pdo->prepare(
            'SELECT id, name, status::text, category, photo_urls, tags FROM pet WHERE id = :id'
        );
        $stmt->execute([':id' => $petId]);
        $row = $stmt->fetch();
        return $row ? $this->hydratePet($row) : null;
    }

    public function updatePetWithForm(int $petId, ?string $name, ?string $status): bool
    {
        $existing = $this->getPetById($petId);
        if ($existing === null) {
            return false;
        }
        if ($name !== null && $name !== '') {
            $existing['name'] = $name;
        }
        if ($status !== null && $status !== '') {
            $existing['status'] = $status;
        }
        $this->addPet($existing);
        return true;
    }

    public function deletePet(int $petId): bool
    {
        $stmt = $this->pdo->prepare('DELETE FROM pet WHERE id = :id');
        $stmt->execute([':id' => $petId]);
        return $stmt->rowCount() > 0;
    }

    public function uploadFile(int $petId, string $content): int
    {
        $sql = <<<SQL
            INSERT INTO pet_photo (pet_id, content)
            VALUES (:pet_id, :content)
            ON CONFLICT (pet_id) DO UPDATE SET content = EXCLUDED.content
        SQL;
        $stmt = $this->pdo->prepare($sql);
        $stmt->bindValue(':pet_id', $petId, PDO::PARAM_INT);
        $stmt->bindValue(':content', $content, PDO::PARAM_LOB);
        $stmt->execute();
        return strlen($content);
    }

    // ── Store ─────────────────────────────────────────────────────────────

    public function getInventory(): array
    {
        $stmt = $this->pdo->query(
            'SELECT status::text AS status, COUNT(*) AS cnt FROM pet GROUP BY status'
        );
        $result = [];
        foreach ($stmt->fetchAll() as $row) {
            $result[$row['status']] = (int) $row['cnt'];
        }
        return $result;
    }

    public function placeOrder(array $order): array
    {
        $id       = $order['id'] ?? null;
        $petId    = $order['petId'] ?? 0;
        $quantity = $order['quantity'] ?? 0;
        $shipDate = $order['shipDate'] ?? null;
        $status   = $order['status'] ?? 'placed';
        $complete = $order['complete'] ?? false;

        if ($id === null || $id === 0) {
            $stmt = $this->pdo->query('SELECT COALESCE(MAX(id),0)+1 AS next_id FROM "order"');
            $id   = (int) $stmt->fetchColumn();
        }

        $sql = <<<SQL
            INSERT INTO "order" (id, pet_id, quantity, ship_date, status, complete)
            VALUES (:id, :pet_id, :quantity, :ship_date, :status::order_status, :complete)
            ON CONFLICT (id) DO UPDATE
                SET pet_id    = EXCLUDED.pet_id,
                    quantity  = EXCLUDED.quantity,
                    ship_date = EXCLUDED.ship_date,
                    status    = EXCLUDED.status,
                    complete  = EXCLUDED.complete
            RETURNING id, pet_id, quantity, ship_date, status::text, complete
        SQL;

        $stmt = $this->pdo->prepare($sql);
        $stmt->execute([
            ':id'        => $id,
            ':pet_id'    => $petId,
            ':quantity'  => $quantity,
            ':ship_date' => $shipDate,
            ':status'    => $status,
            ':complete'  => $complete ? 'true' : 'false',
        ]);

        return $this->hydrateOrder($stmt->fetch());
    }

    public function getOrderById(int $orderId): ?array
    {
        $stmt = $this->pdo->prepare(
            'SELECT id, pet_id, quantity, ship_date, status::text, complete FROM "order" WHERE id = :id'
        );
        $stmt->execute([':id' => $orderId]);
        $row = $stmt->fetch();
        return $row ? $this->hydrateOrder($row) : null;
    }

    public function deleteOrder(int $orderId): bool
    {
        $stmt = $this->pdo->prepare('DELETE FROM "order" WHERE id = :id');
        $stmt->execute([':id' => $orderId]);
        return $stmt->rowCount() > 0;
    }

    // ── User ──────────────────────────────────────────────────────────────

    public function createUser(array $user): array
    {
        $id         = $user['id'] ?? null;
        $username   = $user['username'] ?? '';
        $firstName  = $user['firstName'] ?? null;
        $lastName   = $user['lastName'] ?? null;
        $email      = $user['email'] ?? null;
        $password   = $user['password'] ?? null;
        $phone      = $user['phone'] ?? null;
        $userStatus = $user['userStatus'] ?? 0;

        if ($id === null || $id === 0) {
            $stmt = $this->pdo->query('SELECT COALESCE(MAX(id),0)+1 AS next_id FROM "user"');
            $id   = (int) $stmt->fetchColumn();
        }

        $sql = <<<SQL
            INSERT INTO "user" (id, username, first_name, last_name, email, password, phone, user_status)
            VALUES (:id, :username, :first_name, :last_name, :email, :password, :phone, :user_status)
            ON CONFLICT (id) DO UPDATE
                SET username    = EXCLUDED.username,
                    first_name  = EXCLUDED.first_name,
                    last_name   = EXCLUDED.last_name,
                    email       = EXCLUDED.email,
                    password    = EXCLUDED.password,
                    phone       = EXCLUDED.phone,
                    user_status = EXCLUDED.user_status
            RETURNING id, username, first_name, last_name, email, password, phone, user_status
        SQL;

        $stmt = $this->pdo->prepare($sql);
        $stmt->execute([
            ':id'          => $id,
            ':username'    => $username,
            ':first_name'  => $firstName,
            ':last_name'   => $lastName,
            ':email'       => $email,
            ':password'    => $password,
            ':phone'       => $phone,
            ':user_status' => $userStatus,
        ]);

        return $this->hydrateUser($stmt->fetch());
    }

    public function getUserByName(string $username): ?array
    {
        $stmt = $this->pdo->prepare(
            'SELECT id, username, first_name, last_name, email, password, phone, user_status FROM "user" WHERE username = :username'
        );
        $stmt->execute([':username' => $username]);
        $row = $stmt->fetch();
        return $row ? $this->hydrateUser($row) : null;
    }

    public function updateUser(string $username, array $user): bool
    {
        $existing = $this->getUserByName($username);
        if ($existing === null) {
            return false;
        }
        $user['id']       = $existing['id'];
        $user['username'] = $username;
        $this->createUser($user);
        return true;
    }

    public function deleteUser(string $username): bool
    {
        $stmt = $this->pdo->prepare('DELETE FROM "user" WHERE username = :username');
        $stmt->execute([':username' => $username]);
        return $stmt->rowCount() > 0;
    }

    public function loginUser(string $username, string $password): ?string
    {
        $user = $this->getUserByName($username);
        if ($user === null || ($user['password'] ?? '') !== $password) {
            return null;
        }
        return 'token-' . $username . '-' . time();
    }

    // ── Hydration helpers ─────────────────────────────────────────────────

    private function hydratePet(array $row): array
    {
        return [
            'id'        => (int) $row['id'],
            'name'      => $row['name'],
            'status'    => $row['status'],
            'category'  => isset($row['category']) ? json_decode($row['category'], true) : null,
            'photoUrls' => json_decode($row['photo_urls'] ?? '[]', true),
            'tags'      => json_decode($row['tags'] ?? '[]', true),
        ];
    }

    private function hydrateOrder(array $row): array
    {
        return [
            'id'       => (int) $row['id'],
            'petId'    => (int) $row['pet_id'],
            'quantity' => (int) $row['quantity'],
            'shipDate' => $row['ship_date'],
            'status'   => $row['status'],
            'complete' => (bool) $row['complete'],
        ];
    }

    private function hydrateUser(array $row): array
    {
        return [
            'id'         => (int) $row['id'],
            'username'   => $row['username'],
            'firstName'  => $row['first_name'],
            'lastName'   => $row['last_name'],
            'email'      => $row['email'],
            'password'   => $row['password'],
            'phone'      => $row['phone'],
            'userStatus' => (int) $row['user_status'],
        ];
    }
}

File: petshop-stacks/csharp/aspnetcore/src/Petstore/Repositories/PostgresPetstoreRepository.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using Dapper;
using Newtonsoft.Json;
using Npgsql;
using Petstore.Models;

namespace Petstore.Repositories
{
    public class PostgresPetstoreRepository : IPetstoreRepository
    {
        private readonly NpgsqlDataSource _dataSource;

        public PostgresPetstoreRepository(NpgsqlDataSource dataSource)
        {
            _dataSource = dataSource;
        }

        public static string BuildConnectionString()
        {
            var databaseUrl = Environment.GetEnvironmentVariable("DATABASE_URL");
            if (!string.IsNullOrEmpty(databaseUrl))
                return ApplyPoolSize(databaseUrl);

            var host = Environment.GetEnvironmentVariable("POSTGRES_HOST") ?? "localhost";
            var port = Environment.GetEnvironmentVariable("POSTGRES_PORT") ?? "5434";
            var user = Environment.GetEnvironmentVariable("POSTGRES_USER") ?? "myuser";
            var password = Environment.GetEnvironmentVariable("POSTGRES_PASSWORD") ?? "mypassword";
            var db = Environment.GetEnvironmentVariable("POSTGRES_DB") ?? "csharp-aspnetcore";
            return ApplyPoolSize($"Host={host};Port={port};Username={user};Password={password};Database={db}");
        }

        // Appends "Maximum Pool Size" from PG_MAX_POOL_SIZE when set; unset keeps
        // the Npgsql default (100). Skips appending if the connection string
        // already specifies a pool size, so explicit DATABASE_URL settings win.
        public static string ApplyPoolSize(string connectionString)
        {
            var poolSize = Environment.GetEnvironmentVariable("PG_MAX_POOL_SIZE");
            if (string.IsNullOrEmpty(poolSize))
                return connectionString;

            var lower = connectionString.ToLowerInvariant();
            if (lower.Contains("maximum pool size") || lower.Contains("max pool size") || lower.Contains("maxpoolsize"))
                return connectionString;

            return $"{connectionString};Maximum Pool Size={poolSize}";
        }

        private async Task<NpgsqlConnection> OpenConnectionAsync() =>
            await _dataSource.OpenConnectionAsync();

        // ── Status helpers ─────────────────────────────────────────────

        private static string PetStatusToString(Pet.StatusEnum s) => s switch
        {
            Pet.StatusEnum.AvailableEnum => "available",
            Pet.StatusEnum.PendingEnum   => "pending",
            Pet.StatusEnum.SoldEnum      => "sold",
            _                            => null
        };

        private static Pet.StatusEnum StringToPetStatus(string s) => s switch
        {
            "pending" => Pet.StatusEnum.PendingEnum,
            "sold"    => Pet.StatusEnum.SoldEnum,
            _         => Pet.StatusEnum.AvailableEnum
        };

        private static string OrderStatusToString(Order.StatusEnum s) => s switch
        {
            Order.StatusEnum.ApprovedEnum  => "approved",
            Order.StatusEnum.DeliveredEnum => "delivered",
            Order.StatusEnum.PlacedEnum    => "placed",
            _                              => null
        };

        private static Order.StatusEnum StringToOrderStatus(string s) => s switch
        {
            "approved"  => Order.StatusEnum.ApprovedEnum,
            "delivered" => Order.StatusEnum.DeliveredEnum,
            _           => Order.StatusEnum.PlacedEnum
        };

        // ── Pet row → model ────────────────────────────────────────────

        private static Pet MapPetRow(dynamic row)
        {
            return new Pet
            {
                Id        = row.id,
                Name      = row.name,
                PhotoUrls = row.photo_urls != null
                    ? JsonConvert.DeserializeObject<List<string>>((string)row.photo_urls)
                    : new List<string>(),
                Tags = row.tags != null
                    ? JsonConvert.DeserializeObject<List<Tag>>((string)row.tags)
                    : null,
                Category = row.category != null
                    ? JsonConvert.DeserializeObject<Category>((string)row.category)
                    : null,
                Status = row.status != null
                    ? StringToPetStatus((string)row.status)
                    : default
            };
        }

        // ── Order row → model ──────────────────────────────────────────

        private static Order MapOrderRow(dynamic row)
        {
            return new Order
            {
                Id       = row.id,
                PetId    = row.pet_id,
                Quantity = row.quantity,
                ShipDate = row.ship_date != null ? (DateTime)row.ship_date : default,
                Status   = row.status != null ? StringToOrderStatus((string)row.status) : default,
                Complete = row.complete
            };
        }

        // ── User row → model ───────────────────────────────────────────

        private static User MapUserRow(dynamic row)
        {
            return new User
            {
                Id         = row.id ?? 0L,
                Username   = row.username,
                FirstName  = row.first_name,
                LastName   = row.last_name,
                Email      = row.email,
                Password   = row.password,
                Phone      = row.phone,
                UserStatus = row.user_status ?? 0
            };
        }

        // ══════════════════════ PET OPERATIONS ════════════════════════

        public async Task<Pet> AddPet(Pet pet)
        {
            await using var conn = await OpenConnectionAsync();
            var categoryJson  = pet.Category  != null ? JsonConvert.SerializeObject(pet.Category)  : null;
            var photoUrlsJson = JsonConvert.SerializeObject(pet.PhotoUrls ?? new List<string>());
            var tagsJson      = pet.Tags != null ? JsonConvert.SerializeObject(pet.Tags) : null;
            var statusStr     = PetStatusToString(pet.Status);

            // Single round-trip: COALESCE handles Id=0 (auto-assign via sequence) vs Id>0
            pet.Id = await conn.QuerySingleAsync<long>(
                @"INSERT INTO pet (id, name, category, photo_urls, tags, status)
                  VALUES (COALESCE(NULLIF(@Id, 0), nextval('pet_id_seq')), @Name, @Category::json, @PhotoUrls::json, @Tags::json, @Status::pet_status)
                  ON CONFLICT (id) DO UPDATE SET
                      name       = EXCLUDED.name,
                      category   = EXCLUDED.category,
                      photo_urls = EXCLUDED.photo_urls,
                      tags       = EXCLUDED.tags,
                      status     = EXCLUDED.status
                  RETURNING id",
                new { Id = pet.Id, Name = pet.Name, Category = categoryJson,
                      PhotoUrls = photoUrlsJson, Tags = tagsJson, Status = statusStr });

            return pet;
        }

        public async Task<bool> DeletePet(long petId)
        {
            await using var conn = await OpenConnectionAsync();
            var rows = await conn.ExecuteAsync("DELETE FROM pet WHERE id = @Id", new { Id = petId });
            return rows > 0;
        }

        public async Task<List<Pet>> FindPetsByStatus(string status)
        {
            await using var conn = await OpenConnectionAsync();
            IEnumerable<dynamic> rows;
            if (string.IsNullOrEmpty(status))
            {
                rows = await conn.QueryAsync<dynamic>(
                    "SELECT id, name, category::text, photo_urls::text, tags::text, status::text FROM pet");
            }
            else
            {
                rows = await conn.QueryAsync<dynamic>(
                    "SELECT id, name, category::text, photo_urls::text, tags::text, status::text FROM pet WHERE status::text = @Status",
                    new { Status = status });
            }
            return rows.Select(MapPetRow).ToList();
        }

        public async Task<List<Pet>> FindPetsByTags(List<string> tags)
        {
            await using var conn = await OpenConnectionAsync();

            if (tags == null || tags.Count == 0)
            {
                var allRows = await conn.QueryAsync<dynamic>(
                    "SELECT id, name, category::text, photo_urls::text, tags::text, status::text FROM pet");
                return allRows.Select(MapPetRow).ToList();
            }

            // Push tag filtering to the database using json_array_elements — avoids full table scan in C#
            var rows = await conn.QueryAsync<dynamic>(
                @"SELECT id, name, category::text, photo_urls::text, tags::text, status::text
                  FROM pet
                  WHERE tags IS NOT NULL
                    AND EXISTS (
                        SELECT 1 FROM json_array_elements(tags) elem
                        WHERE elem->>'name' = ANY(@Tags)
                    )",
                new { Tags = tags.ToArray() });

            return rows.Select(MapPetRow).ToList();
        }

        public async Task<Pet> GetPetById(long petId)
        {
            await using var conn = await OpenConnectionAsync();
            var row = await conn.QueryFirstOrDefaultAsync<dynamic>(
                "SELECT id, name, category::text, photo_urls::text, tags::text, status::text FROM pet WHERE id = @Id",
                new { Id = petId });
            return row == null ? null : MapPetRow(row);
        }

        public async Task<Pet> UpdatePet(Pet pet)
        {
            await using var conn = await OpenConnectionAsync();
            var categoryJson  = pet.Category  != null ? JsonConvert.SerializeObject(pet.Category)  : null;
            var photoUrlsJson = JsonConvert.SerializeObject(pet.PhotoUrls ?? new List<string>());
            var tagsJson      = pet.Tags != null ? JsonConvert.SerializeObject(pet.Tags) : null;
            var statusStr     = PetStatusToString(pet.Status);

            // Single round-trip: check affected rows instead of a prior SELECT
            var rows = await conn.ExecuteAsync(
                @"UPDATE pet SET name = @Name, category = @Category::json,
                      photo_urls = @PhotoUrls::json, tags = @Tags::json, status = @Status::pet_status
                  WHERE id = @Id",
                new { Id = pet.Id, Name = pet.Name, Category = categoryJson,
                      PhotoUrls = photoUrlsJson, Tags = tagsJson, Status = statusStr });

            return rows > 0 ? pet : null;
        }

        public async Task<bool> UpdatePetWithForm(long petId, string name, string status)
        {
            await using var conn = await OpenConnectionAsync();

            // Single round-trip: COALESCE keeps existing value when parameter is null
            var rows = await conn.ExecuteAsync(
                @"UPDATE pet SET
                      name   = COALESCE(@Name, name),
                      status = CASE WHEN @Status IS NULL THEN status ELSE @Status::pet_status END
                  WHERE id = @Id",
                new { Id = petId, Name = name, Status = status });

            return rows > 0;
        }

        public async Task<ApiResponse> UploadFile(long petId, string additionalMetadata, Stream fileData)
        {
            byte[] bytes;
            using (var ms = new MemoryStream())
            {
                await fileData.CopyToAsync(ms);
                bytes = ms.ToArray();
            }

            await using var conn = await OpenConnectionAsync();

            // Single round-trip CTE: verifies pet exists, gets next sequence id, and inserts —
            // returns null rows (→ null) when the pet does not exist.
            var photoId = await conn.QuerySingleOrDefaultAsync<long?>(
                @"WITH check_pet AS (
                      SELECT id FROM pet WHERE id = @PetId
                  ),
                  new_id AS (
                      SELECT nextval('pet_photo_id_seq') AS id FROM check_pet
                  )
                  INSERT INTO pet_photo (id, pet_id, metadata, content)
                  SELECT id, @PetId, @Metadata, @Content FROM new_id
                  RETURNING id",
                new { PetId = petId, Metadata = additionalMetadata, Content = bytes });

            if (photoId == null) return null;

            return new ApiResponse
            {
                Code    = 200,
                Message = $"File uploaded, {bytes.Length} bytes stored."
            };
        }

        // ══════════════════════ STORE OPERATIONS ══════════════════════

        public async Task<bool> DeleteOrder(long orderId)
        {
            await using var conn = await OpenConnectionAsync();
            var rows = await conn.ExecuteAsync("DELETE FROM \"order\" WHERE id = @Id", new { Id = orderId });
            return rows > 0;
        }

        public async Task<Dictionary<string, int>> GetInventory()
        {
            await using var conn = await OpenConnectionAsync();
            var rows = await conn.QueryAsync<dynamic>(
                "SELECT status::text AS status, COUNT(*)::int AS cnt FROM pet GROUP BY status");
            var result = new Dictionary<string, int>();
            foreach (var r in rows)
            {
                var key = (string)r.status ?? "unknown";
                result[key] = (int)r.cnt;
            }
            return result;
        }

        public async Task<Order> GetOrderById(long orderId)
        {
            await using var conn = await OpenConnectionAsync();
            var row = await conn.QueryFirstOrDefaultAsync<dynamic>(
                "SELECT id, pet_id, quantity, ship_date, status::text, complete FROM \"order\" WHERE id = @Id",
                new { Id = orderId });
            return row == null ? null : MapOrderRow(row);
        }

        public async Task<Order> PlaceOrder(Order order)
        {
            await using var conn = await OpenConnectionAsync();
            var statusStr = OrderStatusToString(order.Status);

            // Single round-trip: COALESCE handles Id=0 (auto-assign via sequence) vs Id>0
            order.Id = await conn.QuerySingleAsync<long>(
                @"INSERT INTO ""order"" (id, pet_id, quantity, ship_date, status, complete)
                  VALUES (COALESCE(NULLIF(@Id, 0), nextval('order_id_seq')), @PetId, @Quantity, @ShipDate, @Status::order_status, @Complete)
                  ON CONFLICT (id) DO UPDATE SET
                      pet_id    = EXCLUDED.pet_id,
                      quantity  = EXCLUDED.quantity,
                      ship_date = EXCLUDED.ship_date,
                      status    = EXCLUDED.status,
                      complete  = EXCLUDED.complete
                  RETURNING id",
                new { Id = order.Id, PetId = order.PetId, order.Quantity,
                      ShipDate = order.ShipDate == default ? (DateTime?)null : order.ShipDate,
                      Status = statusStr, order.Complete });

            return order;
        }

        // ══════════════════════ USER OPERATIONS ═══════════════════════

        public async Task<User> CreateUser(User user)
        {
            await using var conn = await OpenConnectionAsync();
            return await InsertUserAsync(conn, user);
        }

        public async Task<User> CreateUsersWithListInput(List<User> users)
        {
            if (users == null || users.Count == 0)
                return null;

            // Single connection shared across all inserts — avoids N connection acquisitions
            await using var conn = await OpenConnectionAsync();
            User last = null;
            foreach (var u in users)
                last = await InsertUserAsync(conn, u);
            return last;
        }

        private static async Task<User> InsertUserAsync(NpgsqlConnection conn, User user)
        {
            // Single round-trip: COALESCE handles Id=0 (auto-assign via sequence) vs Id>0
            user.Id = await conn.QuerySingleAsync<long>(
                @"INSERT INTO ""user"" (id, username, first_name, last_name, email, password, phone, user_status)
                  VALUES (COALESCE(NULLIF(@Id, 0), nextval('user_id_seq')), @Username, @FirstName, @LastName, @Email, @Password, @Phone, @UserStatus)
                  ON CONFLICT (username) DO UPDATE SET
                      id          = EXCLUDED.id,
                      first_name  = EXCLUDED.first_name,
                      last_name   = EXCLUDED.last_name,
                      email       = EXCLUDED.email,
                      password    = EXCLUDED.password,
                      phone       = EXCLUDED.phone,
                      user_status = EXCLUDED.user_status
                  RETURNING id",
                new { user.Id, user.Username, FirstName = user.FirstName,
                      LastName = user.LastName, user.Email, user.Password,
                      user.Phone, UserStatus = user.UserStatus });
            return user;
        }

        public async Task<bool> DeleteUser(string username)
        {
            await using var conn = await OpenConnectionAsync();
            var rows = await conn.ExecuteAsync(
                "DELETE FROM \"user\" WHERE username = @Username", new { Username = username });
            return rows > 0;
        }

        public async Task<User> GetUserByName(string username)
        {
            await using var conn = await OpenConnectionAsync();
            var row = await conn.QueryFirstOrDefaultAsync<dynamic>(
                "SELECT id, username, first_name, last_name, email, password, phone, user_status FROM \"user\" WHERE username = @Username",
                new { Username = username });
            return row == null ? null : MapUserRow(row);
        }

        public Task<string> LoginUser(string username, string password)
        {
            return Task.FromResult("logged-in");
        }

        public Task LogoutUser()
        {
            return Task.CompletedTask;
        }

        public async Task<bool> UpdateUser(string username, User user)
        {
            await using var conn = await OpenConnectionAsync();
            var rows = await conn.ExecuteAsync(
                @"UPDATE ""user"" SET first_name = @FirstName, last_name = @LastName,
                      email = @Email, password = @Password, phone = @Phone, user_status = @UserStatus
                  WHERE username = @Username",
                new { Username = username, FirstName = user.FirstName,
                      LastName = user.LastName, user.Email, user.Password,
                      user.Phone, UserStatus = user.UserStatus });
            return rows > 0;
        }
    }
}
